src/realm/object-store/sync/sync_session.cpp

```diff
@@ -326,14 +378,18 @@ void SyncSession::handle_fresh_realm_downloaded(DBRef db, util::Optional<std::st
     // re-registered when the session becomes active again.
     {
         m_force_client_reset = true;
-        m_client_reset_fresh_copy = std::move(db);
+        m_client_reset_fresh_copy = db;
         CompletionCallbacks callbacks;
         std::swap(m_completion_callbacks, callbacks);
         // always swap back, even if advance_state throws
         auto guard = util::make_scope_exit([&]() noexcept {
-            std::swap(callbacks, m_completion_callbacks);
+            util::CheckedUniqueLock lock(m_state_mutex);
```

Comment:
This previously was a data race if the user tried to add a completion callback at exactly the wrong time. We also don't want to discard any completion callbacks added during the brief window where we didn't hold the mutex.